skyportal/facility_apis/observation_plan.py

```diff
@@ -0,0 +1,337 @@
+
+        params['is3D'] = request.localization.is_3d
+
+        # Function to read maps
+        map_struct = gwemopt.utils.read_skymap(
+            params, is3D=params["do3D"], map_struct=params['map_struct']
+        )
+
+        moc_structs = gwemopt.skyportal.create_moc_from_skyportal(
+            params, map_struct=map_struct
+        )
+        tile_structs = gwemopt.tiles.moc(params, map_struct, moc_structs)
+
+        tile_structs, coverage_struct = gwemopt.coverage.timeallocation(
+            params, map_struct, tile_structs
+        )
+
+        planned_observations = []
+        for ii in range(len(coverage_struct["ipix"])):
+            data = coverage_struct["data"][ii, :]
+            filt = coverage_struct["filters"][ii]
+            mjd = data[2]
+            tt = Time(mjd, format='mjd')
+
+            overhead_per_exposure = params["config"][request.instrument.name][
+                "overhead_per_exposure"
+            ]
+
+            exposure_time, field_id, prob = data[4], data[5], data[6]
+
+            field = InstrumentField.query.filter(
+                InstrumentField.instrument_id == request.instrument.id,
+                InstrumentField.id == field_id,
+            ).first()
+            if field is None:
+                return log(f"Missing field {field_id} from list")
+
+            planned_observation = PlannedObservation(
+                obstime=tt.datetime,
+                dateobs=request.gcnevent.dateobs,
+                field_id=field.id,
+                exposure_time=exposure_time,
+                weight=prob,
+                filt=filt,
+                instrument_id=request.instrument.id,
+                planned_observation_id=ii,
+                observation_plan_id=plan.id,
+                overhead_per_exposure=overhead_per_exposure,
+            )
+            planned_observations.append(planned_observation)
+
+        session.add_all(planned_observations)
+        plan.status = 'complete'
+        session.merge(plan)
+        session.commit()
+
+        request.status = 'complete'
+        session.merge(request)
+        session.commit()
+
+        flow = Flow()
+        flow.push(
+            '*',
+            "skyportal/REFRESH_GCNEVENT",
+            payload={"gcnEvent_dateobs": request.gcnevent.dateobs},
+        )
+
+        return log(f"Generated plan for observation plan {observation_plan_id}")
+
+    except Exception as e:
+        return log(
+            f"Unable to generate plan for observation plan {observation_plan_id}: {e}"
+        )
+    finally:
+        Session.remove()
+
+
+class MMAAPI(FollowUpAPI):
+
+    """An interface to MMA operations."""
+
+    # subclasses *must* implement the method below
+    @staticmethod
+    def submit(request):
+
+        """Generate an observation plan.
+
+        Parameters
+        ----------
+        request: skyportal.models.ObservationPlanRequest
+            The request to generate the observation plan and the SkyPortal database.
```

Comment:
"on the database"? or maybe I didn't understand...

Reply:
lolz